go/new_merkletree/merkle_tree.go

```diff
@@ -0,0 +1,177 @@
+package new_merkletree
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"hash"
+	"math"
+)
+
+// New creates a new merkle hash tree.  The number of "leaves" of the tree,
+// as well as their contents, are retrieved through |dao|.  If you wish to
+// have acceptable performance on non-trivial tree sizes, you'll want to
+// provide |cache| (otherwise, pass `nil`).  The hash function used for all
+// nodes in the tree is specified by |hasher|.
+func New(dao NewMerkleTreeDataInterface, cache NewMerkleTreeCacheInterface, hasher hash.Hash) *NewMerkleTree {
+	return &NewMerkleTree{dao: dao, cache: cache, hasher: hasher}
+}
+
+func (mt NewMerkleTree) CurrentRoot() (Hash, error) {
+	if mt.dao.Size() == 0 {
+		// Special case: empty trees get hashes of empty strings
+		mt.hasher.Reset()
+		return mt.hasher.Sum([]byte{}), nil
+	} else {
+		return mt.subtreeRoot(0, mt.dao.Size()-1)
+	}
+}
+
+func (mt NewMerkleTree) InclusionProof(leaf uint64) ([]Hash, error) {
+	if mt.dao.Size() == 0 {
+		return nil, errors.New("NewMerkleTree: Can't calculate an inclusion proof on an empty tree")
+	}
+
+	if leaf >= mt.dao.Size() {
+		return nil, fmt.Errorf("NewMerkleTree: Invalid leaf entry ID: %v", leaf)
+	}
+
+	return mt.inclusionSubtree(leaf, 0, mt.dao.Size()-1)
+}
+
+func (mt NewMerkleTree) ConsistencyProof(from, to uint64) ([]Hash, error) {
+	if from == 0 {
```

Comment:
This might be a good candidate for Go's switch statement:
```
switch {
       from == 0:
               ...
               return []Hash{}, nil
        to > mt.dao.Size():
               ...
               return ...
        from > to:
               ...
}
return mt.subproof(...)